core/src/ephemeral_server/mod.rs

```diff
@@ -0,0 +1,461 @@
+                };
+                // Create dest file based on SDK name/version or fixed version
+                let dest = dest_dir.join(match version {
+                    EphemeralExeVersion::Default {
+                        sdk_name,
+                        sdk_version,
+                    } => format!("{}-{}-{}{}", artifact_name, sdk_name, sdk_version, out_ext),
+                    EphemeralExeVersion::Fixed(version) => {
+                        format!("{}-{}{}", artifact_name, version, out_ext)
+                    }
+                });
+                debug!(
+                    "Lazily downloading or using existing exe at {}",
+                    dest.display()
+                );
+
+                // If it already exists, skip
+                if dest.exists() {
+                    return Ok(dest);
+                }
+
+                // Get info about the proper archive and in-archive file
+                let arch = match std::env::consts::ARCH {
+                    "x86_64" => "amd64",
+                    "arm" | "aarch64" => "arm64",
+                    other => return Err(anyhow!("Unsupported arch: {}", other)),
+                };
+                let mut get_info_params = vec![("arch", arch), ("platform", platform)];
+                let version_name = match version {
+                    EphemeralExeVersion::Default {
+                        sdk_name,
+                        sdk_version,
+                    } => {
+                        get_info_params.push(("sdk-name", sdk_name.as_str()));
+                        get_info_params.push(("sdk-version", sdk_version.as_str()));
+                        "default"
+                    }
+                    EphemeralExeVersion::Fixed(version) => version,
+                };
+                let client = reqwest::Client::new();
+                let info: DownloadInfo = client
+                    .get(format!(
+                        "https://temporal.download/{}/{}",
+                        artifact_name, version_name
+                    ))
+                    .query(&get_info_params)
+                    .send()
+                    .await?
+                    .json()
+                    .await?;
+
+                // Attempt download, looping because it could have waited for
+                // concurrent one to finish
+                loop {
+                    if lazy_download_exe(
+                        &client,
+                        &info.archive_url,
+                        Path::new(&info.file_to_extract),
+                        &dest,
+                    )
+                    .await?
+                    {
+                        return Ok(dest);
+                    }
+                }
+            }
+        }
+    }
+}
+
+fn get_free_port() -> u16 {
+    // Can just ask OS to give us a port then close socket. OS's don't give that
+    // port back to anyone else anytime soon.
+    std::net::TcpListener::bind("127.0.0.1:0")
```

Comment:
You should use the IP address provided in the config

Reply:
This is just for obtaining a free port from the current system. The user-defined IP is not needed from my understanding of how OS's allocate ports. I guess I could use `0.0.0.0` but I fear locked down Linux capabilities of doing that.

Reply:
We aren't given a port for Java test server. I will pass in the user-supplied one for Temporalite and use 0.0.0.0 for Java test server.

Reply:
> We aren't given a port for Java test server

If we support user defined port (arguably not needed) for temporalite, we should for Java test server as well.

Reply:
Sorry, I meant we aren't given an _ip_

Reply:
Added port for free port obtaining